src/autoscaler-common/counters_base.js

```diff
@@ -0,0 +1,448 @@
+ * number of errors reported.
+ *
+ * @param {Object} err
+ */
+function openTelemetryGlobalErrorHandler(err) {
+  openTelemetryErrorCount++;
+  // delegate to Otel's own error handler for stringification
+  OpenTelemetryCore.loggingErrorHandler()(err);
+}
+
+
+// Setup OpenTelemetry client libraries logging.
+OpenTelemetryApi.default.diag.setLogger(
+    DIAG_BUNYAN_LOGGER,
+    {
+      logLevel: OpenTelemetryApi.DiagLogLevel.INFO,
+      suppressOverrideMessage: true,
+    });
+OpenTelemetryCore.setGlobalErrorHandler(openTelemetryGlobalErrorHandler);
+
+
+/**
+ * Initialize the OpenTelemetry metrics, set up logging
+ *
+ * If called more than once, will wait for the first call to complete.
+ *
+ * @return {!Promise<void>}
+ */
+async function initMetrics() {
+  // check to see if someone else has started to init counters before
+  // so that this function only runs once.
+  if (pendingInit) {
+    return await pendingInit;
+  }
+
+  /** @type {?function():void} */
+  let resolvePendingInit = null;
+  /** @type {?function(?):void} */
+  let rejectPendingInit = null;
+  pendingInit = new Promise((res, rej) => {
+    resolvePendingInit = res;
+    rejectPendingInit = rej;
+  });
+
+
+  try {
+    logger.debug('initializing metrics');
+
+    if (process.env.KUBERNETES_SERVICE_HOST) {
+      // In K8s. We need to set the Pod Name to prevent duplicate
+      // timeseries errors.
+      if (process.env.K8S_POD_NAME) {
+        RESOURCE_ATTRIBUTES[
+            Semconv.K8S_POD_NAME] =
+            process.env.K8S_POD_NAME;
+      } else {
+        logger.warn('WARNING: running under Kubernetes, but K8S_POD_NAME ' +
+        'environment variable is not set. ' +
+        'This may lead to Send TimeSeries errors');
+      }
+    }
+
+    const gcpResources = new GcpDetectorSync().detect();
+    await gcpResources.waitForAsyncAttributes();
+    logger.debug('Got gcpResources attrs:1 %o', gcpResources.attributes);
+
+    if (process.env.FUNCTION_TARGET) {
+      // In cloud functions.
+      // We need to set the platform to generic_task so that the
+      // function instance ID gets set in the  counter resource attributes.
+      // For details, see
+      // https://github.com/GoogleCloudPlatform/opentelemetry-operations-js/issues/679
+      RESOURCE_ATTRIBUTES[Semconv.CLOUD_PLATFORM] = 'generic_task';
+
+      if (gcpResources.attributes[Semconv.FAAS_ID]?.toString()) {
+        RESOURCE_ATTRIBUTES[Semconv.SERVICE_INSTANCE_ID] =
+            gcpResources.attributes[Semconv.FAAS_ID].toString();
+      } else {
+        logger.warn('WARNING: running under Cloud Functions, but FAAS_ID ' +
+        'resource attribute is not set. ' +
+        'This may lead to Send TimeSeries errors');
+      }
+    }
+
+    const resources = gcpResources.merge(new Resource(RESOURCE_ATTRIBUTES));
+    await resources.waitForAsyncAttributes();
+
+    logger.debug('Got metrics resource attrs2: %o', resources.attributes);
```

Comment:
Nit: the formatting of the object in the debug message body takes up multiline real-estate in the top-level logging views in the web console -- do you think it's worth embedding this in a different field?

Reply:
removed